File: csharp/src/Vista.SDK.System.Text.Json/TimeSeriesData/Extensions.cs
using Domain = Vista.SDK.Transport.TimeSeries;

namespace Vista.SDK.Transport.Json.TimeSeriesData;

public static class Extensions
{
    public static TimeSeriesDataPackage ToJsonDto(this Domain.TimeSeriesDataPackage package)
    {
        var p = package.Package;
        var h = package.Package.Header;
        return new TimeSeriesDataPackage(
            new Package(
                h is null
                    ? null
                    : new Header(
                        h.Author,
                        h.DateCreated,
                        h.DateModified,
                        h.ShipId.ToString(),
                        h.SystemConfiguration is null
                            ? null
                            : h.SystemConfiguration.Select(r => new ConfigurationReference(r.Id, r.TimeStamp)).ToList(),
                        h.TimeSpan is null ? null : new TimeSpan(h.TimeSpan.End, h.TimeSpan.Start)
                    )
                    {
                        CustomHeaders = h.CustomHeaders?.CopyProperties()
                    },
                p.TimeSeriesData
                    .Select(
                        t =>
                            new TimeSeriesData(
                                t.DataConfiguration is null
                                    ? null
                                    : new ConfigurationReference(t.DataConfiguration.Id, t.DataConfiguration.TimeStamp),
                                t.EventData is null
                                    ? null
                                    : new EventData(
                                        t.EventData
                                            .DataSet
                                            ?.Select(
                                                d =>
                                                    new DataSet_Event(
                                                        d.DataChannelId.ToString(),
                                                        d.Quality,
                                                        d.TimeStamp,
                                                        d.Value
                                                    )
                                            )
                                            .ToList(),
                                        t.EventData.NumberOfDataSet
                                    ),
                                t.TabularData
                                    ?.Select(
                                        d =>
                                            new TabularData(
                                                d.DataChannelIds?.Select(i => i.ToString()).ToList(),
                                                d.DataSets
                                                    ?.Select(
                                                        td => new DataSet_Tabular(td.Quality, td.TimeStamp, td.Value)
                                                    )
                                                    .ToList(),
                                                d.NumberOfDataChannels,
                                                d.NumberOfDataSets
                                            )
                                    )
                                    .ToList()
                            )
                            {
                                CustomData = t.CustomDataKinds?.CopyProperties()
                            }
                    )
                    .ToList()
            )
        );
    }

    public static Domain.TimeSeriesDataPackage ToDomainModel(this TimeSeriesDataPackage package)
    {
        var p = package.Package;
        var h = package.Package.Header;
        return new Domain.TimeSeriesDataPackage
        {
            Package = new Domain.Package
            {
                Header = h is null
                    ? null
                    : new Domain.Header
                    {
                        ShipId = ShipId.Parse(h.ShipID),
                        TimeSpan = h.TimeSpan is null
                            ? null
                            : new Domain.TimeSpan { Start = h.TimeSpan.Start, End = h.TimeSpan.End },
                        DateCreated = h.DateCreated,
                        DateModified = h.DateModified,
                        Author = h.Author,
                        SystemConfiguration = h.SystemConfiguration
                            ?.Select(c => new Domain.ConfigurationReference { Id = c.ID, TimeStamp = c.TimeStamp })
                            .ToList(),
                        CustomHeaders = h.CustomHeaders?.CopyProperties()
                    },
                TimeSeriesData = p.TimeSeriesData
                    .Select(
                        t =>
                            new Domain.TimeSeriesData
                            {
                                DataConfiguration = t.DataConfiguration is null
                                    ? null
                                    : new Domain.ConfigurationReference
                                    {
                                        Id = t.DataConfiguration.ID,
                                        TimeStamp = t.DataConfiguration.TimeStamp
                                    },
                                TabularData = t.TabularData
                                    ?.Select(td =>
                                    {
                                        if (td.NumberOfDataChannel != td.DataChannelID?.Count)
                                            throw new ArgumentException(
                                                "Number of data channels does not match the expected count"
                                            );
                                        if (td.NumberOfDataSet != td.DataSet?.Count)
                                            throw new ArgumentException(
                                                "Number of data sets does not match the expected count"
                                            );
                                        return new Domain.TabularData
                                        {
                                            DataChannelIds = td.DataChannelID
                                                ?.Select(i => DataChannelId.Parse(i))
                                                .ToList(),
                                            DataSets = td.DataSet
                                                ?.Select(
                                                    tds =>
                                                        new Domain.TabularDataSet
                                                        {
                                                            TimeStamp = tds.TimeStamp,
                                                            Value = tds.Value.ToList(),
                                                            Quality = tds.Quality?.ToList()
                                                        }
                                                )
                                                .ToList()
                                        };
                                    })
                                    .ToList(),
                                EventData = t.EventData is null
                                    ? null
                                    : new Domain.EventData
                                    {
                                        DataSet = t.EventData
                                            .DataSet
                                            ?.Select(
                                                ed =>
                                                    new Domain.EventDataSet
                                                    {
                                                        TimeStamp = ed.TimeStamp,
                                                        DataChannelId = DataChannelId.Parse(ed.DataChannelID),
                                                        Value = ed.Value,
                                                        Quality = ed.Quality
                                                    }
                                            )
                                            .ToList()
                                    },
                                CustomDataKinds = t.CustomData?.CopyProperties()
                            }
                    )
                    .ToList()
            }
        };
    }
}

File: js/lib/transport/json/time-series-data/Extensions.ts
import { DataChannelId, ShipId, TimeSeries } from "../../domain";
import { TimeSeriesDto } from "./TimeSeriesData";

export class Extensions {
    public static toJsonDto(
        domain: TimeSeries.TimeSeriesDataPackage
    ): TimeSeriesDto.TimeSeriesDataPackage {
        const p = domain.package;
        const h = domain.package.header;

        return {
            Package: {
                Header: h
                    ? {
                          Author: h.author,
                          DateCreated: h.dateCreated,
                          DateModified: h.dateModified,
                          ShipID: h.shipId.toString(),
                          SystemConfiguration: h.systemConfiguration?.map(
                              (s) => ({
                                  ID: s.id,
                                  TimeStamp: s.timeStamp,
                              })
                          ),
                          TimeSpan: h.timeSpan
                              ? { End: h.timeSpan.end, Start: h.timeSpan.start }
                              : undefined,
                          ...h.customHeaders,
                      }
                    : undefined,
                TimeSeriesData: p.timeSeriesData.map((t) => ({
                    DataConfiguration: t.dataConfiguration
                        ? {
                              TimeStamp: t.dataConfiguration.timeStamp,
                              ID: t.dataConfiguration.id,
                          }
                        : undefined,
                    EventData: t.eventData
                        ? {
                              NumberOfDataSet: t.eventData.numberOfDataSet,
                              DataSet: t.eventData.dataSet?.map((d) => ({
                                  TimeStamp: d.timeStamp,
                                  DataChannelID: d.dataChannelId.toString(),
                                  Value: d.value,
                                  Quality: d.quality,
                              })),
                          }
                        : undefined,
                    TabularData: t.tabularData?.map((d) => ({
                        NumberOfDataSet: d.numberOfDataSet,
                        NumberOfDataChannel: d.numberOfDataChannel,
                        DataChannelID: d.dataChannelId?.map((i) =>
                            i.toString()
                        ),
                        DataSet: d.dataSet?.map((td) => ({
                            Quality: td.quality,
                            TimeStamp: td.timeStamp,
                            Value: td.value,
                        })),
                    })),
                    ...t.customProperties,
                })),
            },
        };
    }

    public static async toDomainModel(
        dto: TimeSeriesDto.TimeSeriesDataPackage
    ): Promise<TimeSeries.TimeSeriesDataPackage> {
        const timeSeriesData: TimeSeries.TimeSeriesData[] = [];

        for (let {
            DataConfiguration,
            EventData,
            TabularData,
            ...customProperties
        } of dto.Package.TimeSeriesData) {
            const eventDataSet: TimeSeries.EventDataSet[] = [];
            const tabularData: TimeSeries.TabularData[] = [];

            if (EventData?.DataSet) {
                for (let ed of EventData.DataSet) {
                    const dataChannelId = await DataChannelId.parseAsync(
                        ed.DataChannelID
                    );
                    eventDataSet.push({
                        dataChannelId,
                        timeStamp: ed.TimeStamp,
                        value: ed.Value,
                        quality: ed.Quality,
                    });
                }
            }

            if (TabularData) {
                for (let td of TabularData) {
                    const dataChannelIds: DataChannelId[] = [];
                    for (let i of td.DataChannelID ?? []) {
                        const dataChannelId = await DataChannelId.parseAsync(i);
                        dataChannelIds.push(dataChannelId);
                    }

                    tabularData.push({
                        numberOfDataSet: td.NumberOfDataSet,
                        numberOfDataChannel: td.NumberOfDataChannel,
                        dataChannelId:
                            dataChannelIds.length > 0
                                ? dataChannelIds
                                : undefined,
                        dataSet: td.DataSet?.map((d) => ({
                            timeStamp: d.TimeStamp,
                            value: d.Value,
                            quality: d.Quality,
                        })),
                    });
                }
            }

            timeSeriesData.push({
                dataConfiguration: DataConfiguration
                    ? {
                          timeStamp: DataConfiguration.TimeStamp,
                          id: DataConfiguration.ID,
                      }
                    : undefined,
                eventData: {
                    dataSet: eventDataSet.length > 0 ? eventDataSet : undefined,
                    numberOfDataSet: EventData?.NumberOfDataSet,
                },
                tabularData: tabularData.length > 0 ? tabularData : undefined,
                customProperties: customProperties,
            });
        }

        const mapHeader = (): TimeSeries.Header | undefined => {
            if (!dto.Package.Header) return undefined;

            const {
                ShipID,
                TimeSpan,
                DateCreated,
                DateModified,
                Author,
                SystemConfiguration,
                ...custom
            } = dto.Package.Header;
            return {
                shipId: ShipId.parse(ShipID),
                timeSpan: TimeSpan
                    ? { end: TimeSpan.End, start: TimeSpan.Start }
                    : undefined,
                dateCreated: DateCreated,
                dateModified: DateModified,
                author: Author,
                systemConfiguration: SystemConfiguration?.map((c) => ({
                    id: c.ID,
                    timeStamp: c.TimeStamp,
                })),
                customHeaders: custom,
            };
        };

        return {
            package: {
                header: mapHeader(),
                timeSeriesData: timeSeriesData,
            },
        };
    }
}
